Implement config checking

veracioux opened the pull request:
Tracks issue #1740.
@borgmanJeremy I have completely redesigned `ConfigHandler` and added a class `ValueHandler` (`src/utils/valuehandler.{h,cpp}`). I think it is more maintainable this way, and works better with the config checking mechanism.
In essence, those are the only substantial changes I made. All the other files are mostly a result of changing the names of the config value getters/setters in `ConfigHandler`.

NOTE: This is a draft, but will leave it open for checks to run on it.

## Tasks
### TODO Checking API
- [X] Unrecognized config options
- [X] Shortcut conflicts (multiple shortcuts with the same binding)
#### - Semantic checks
Whether the setting values have the correct format (number, string, bool, etc.)
- [X] boolean
- [X] color
- [X] color list
- [X] existing directory
- [X] button list (`buttons` option)
- [X] bounded integers
- [X] filename pattern
- [x] key shortcuts
### TODO Checking procedure
- [X] On daemon launch
- [X] Config changed while daemon running
       Consider another point from mmahmoudian which I don't really understand.
### TODO Error reporting
- [X] Notification
- [X] Log file
       For this one I added a TODO comment. We should wait for logging to be implemented properly, and then implement config error logging in another PR.
- [X] GUI overlay
       @mmahmoudian and I agreed on a message box before a capture but I thought your original proposal is a better solution. The message appears in the bottom left corner and it automatically disappears if the mouse is over it and a tool is active.
- [X] Config menu indicator
       @mmahmoudian's proposal for a check again button is unnecessary because the
       config is watched for changes
- [x] Reporting the exact error

### TODO Additional things

- [x] remove obsolete config entries
- [x] fall back to paths `~/Pictures`, `~`, `/tmp` in that orders
## Testing
Please test the error checking by introducing errors in the config for various config options and value types. You should get notified every time the error state changes.
- Whenever you introduce an error in the config, the change should be detected and get notified. You will get the notification only the first time an error occurs.
- If you resolve the error, you will get notified as well
- Now if you introduce an error again, you will get notified. And so on...
- If you have another monitor like me, you might be able to edit the config file while the the capture GUI is active to see what happens with the bottom left message. You should get a message that says that the config was resolved. If that isn't possible, then you could create two files - one correct and one erroneous - and a script that applies them to the active config with a timer.
- A special case is when you delete the config file. I had to bypass Qt's `QFileSystemWatcher` behavior here, so this case should be more thoroughly tested.
- I think this should be tested on multiple platforms just in case. I got Linux covered, and will eventually do Windows too.

veracioux commented on src/utils/confighandler.cpp:
```diff
@@ -4,380 +4,175 @@
+    OPTION("disabledTrayIcon"            ,Bool               ( false         )),
+    OPTION("historyConfirmationToDelete" ,Bool               ( true          )),
+    OPTION("checkForUpdates"             ,Bool               ( true          )),
+#if defined(Q_OS_MACOS)
+    CUSTOM("startupLaunch"               ,Bool               ( false         )),
+#else
+    OPTION("startupLaunch"               ,Bool               ( true          )),
+#endif
+    OPTION("showStartupLaunchMessage"    ,Bool               ( true          )),
+    OPTION("copyAndCloseAfterUpload"     ,Bool               ( true          )),
+    OPTION("copyPathAfterSave"           ,Bool               ( false         )),
+#if !defined(Q_OS_MACOS) // TODO is this the right way?
+    OPTION("useJpgForClipboard"          ,Bool               ( false         )),
+#endif
+    // TODO obsolete?
+    OPTION("saveAfterCopy"               ,Bool               ( false         )),
+    OPTION("savePath"                    ,ExistingDir        (               )),
+    OPTION("savePathFixed"               ,Bool               ( false         )),
+    OPTION("uploadHistoryMax"            ,LowerBoundedInt(1  , 25            )),
+    OPTION("undoLimit"                   ,BoundedInt(1, 999  , 100           )),
+    // Interface tab
+    OPTION("uiColor"                     ,Color              ( {116, 0, 150} )),
+    OPTION("contrastUiColor"             ,Color              ( {39, 0, 50}   )),
+    OPTION("contrastOpacity"             ,BoundedInt(0, 255  , 190           )),
+    OPTION("buttons"                     ,ButtonList         ( {}            )),
+    // Filename Editor tab
+    OPTION("filenamePattern"             ,FilenamePattern    ( {}            )),
+    // Others
+    // TODO obsolete?
+    OPTION("saveAfterCopyPath"           ,ExistingDir        (               )),
+    OPTION("drawThickness"               ,LowerBoundedInt(1  , 3             )),
+    OPTION("drawColor"                   ,Color              ( Qt::red       )),
+    OPTION("userColors"                  ,UserColors         (               )),
+    OPTION("drawFontSize"                ,LowerBoundedInt(1  , 8             )),
+    OPTION("ignoreUpdateToVersion"       ,String             ( ""            )),
+    OPTION("keepOpenAppLauncher"         ,Bool               ( false         )),
+    OPTION("fontFamily"                  ,String             ( ""            )),
+    OPTION("setSaveAsFileExtension"      ,String             ( ""            )),
+      // clang-format on
+  };
```
Please review this even if you are not a programmer to see if I have messed up some default values.

veracioux replied:
Great idea!

mmahmoudian commented:
@veracioux I haven't got to the end of it, but while reading "TODO Checking API" I realized we should also check if same shortcut is defined multiple times (regardless of having the same binding or not). For example if we have the following case in the `flameshot.ini`.

> ```
> ...
> [Shortcuts]
> TYPE_ARROW=A
> TYPE_ARROW=B
> TYPE_ARROW=A
> ...
> ```

veracioux commented:
@mmahmoudian Addressed that in **Notes about QSettings**.

veracioux commented:
@mmahmoudian Sorry about the failing builds. Unless you use ubuntu 18.04, you should be good to test the program now. I'm going to have to setup a VM to resolve issues on 18.04.

mmahmoudian commented:
I downloaded the [AppImage](https://github.com/flameshot-org/flameshot/pull/1859/checks?check_run_id=3475343376) which was built against 90541b441deeecba2e60648ff684a301f1d0e80d and it says that my config has error:

![image](https://user-images.githubusercontent.com/390889/131543860-efbf5b53-61e4-425e-810e-542f2fb23235.png)

I don't understand what is the issue, nothing is reported to the stdout and there is no log file:

![image](https://user-images.githubusercontent.com/390889/131544682-db8852d8-ea83-43e1-aef7-6128a3d14f57.png)


<details>
<summary>Here is my config file</summary>

> ```
> [General]
> contrastOpacity=188
> disabledTrayIcon=false
> drawColor=#800080
> drawFontSize=10
> drawThickness=6
> saveAfterCopyPath=/home/mehrad/Pictures
> savePath=/home/mehrad/
> savePathFixed=false
> setSaveAsFileExtension=.png
> showStartupLaunchMessage=true
> startupLaunch=true
> uiColor=#ff7300
> undoLimit=100
> uploadHistoryMax=25
> useJpgForClipboard=false
> 
> [Shortcuts]
> TYPE_ARROW=A
> TYPE_CIRCLE=
> TYPE_CIRCLECOUNT=
> TYPE_COMMIT_CURRENT_TOOL=Ctrl+Return
> TYPE_COPY=Ctrl+C
> TYPE_DELETE_CURRENT_TOOL=Del
> TYPE_DRAWER=D
> TYPE_EXIT=Ctrl+Q
> TYPE_IMAGEUPLOADER=
> TYPE_MARKER=M
> TYPE_MOVESELECTION=Ctrl+M
> TYPE_MOVE_DOWN=Down
> TYPE_MOVE_LEFT=Left
> TYPE_MOVE_RIGHT=Right
> TYPE_MOVE_UP=Up
> TYPE_OPEN_APP=Ctrl+O
> TYPE_PENCIL=
> TYPE_PIN='
> TYPE_PIXELATE=B
> TYPE_RECTANGLE=R
> TYPE_REDO=Ctrl+Shift+Z
> TYPE_RESIZE_DOWN=Shift+Down
> TYPE_RESIZE_LEFT=Shift+Left
> TYPE_RESIZE_RIGHT=Shift+Right
> TYPE_RESIZE_UP=Shift+Up
> TYPE_SAVE=Ctrl+S
> TYPE_SELECTION=S
> TYPE_SELECTIONINDICATOR=
> TYPE_SELECT_ALL=Ctrl+A
> TYPE_TEXT=T
> TYPE_TOGGLE_PANEL=Space
> TYPE_UNDO=Ctrl+Z
> ```

</details>

veracioux commented:
@mmahmoudian There was a bug, should be fixed now. If you get an error again, check if the paths in your config are existing directories.

borgmanJeremy commented:
I have an issue if I try to remap a key via the gui: 
![Peek 2021-09-01 06-59](https://user-images.githubusercontent.com/46930769/131667685-91bac03e-095d-4d1c-b0e6-1991ade25e86.gif)

veracioux commented:
@borgmanJeremy Are you compiling in debug mode? If you are not, could you do that and tell me if you are getting an exception when trying to change a shortcut?

borgmanJeremy commented:
@veracioux  Since there is a bunch of discussion on this topic I thought we could try github "discussions" I posted my reply here: https://github.com/flameshot-org/flameshot/discussions/1862

mmahmoudian commented:
I build from source against b6641ea commit and:

- having `drawFontSize=A` didn't raise an error
- having `uploadHistoryMax=2.5` (with decimal points) would automagically correct it to `uploadHistoryMax=0` which should not become 0
- having`TYPE_RESIZE_RIGHT=Shift+Right:` (extra colon at the end) would not generate error, it is not automagically fixed and the shortcut does not work anymore. even the colon shows up in the config window:
![image](https://user-images.githubusercontent.com/390889/131871897-55ebdba5-cd54-4b46-b189-1fed15a08b50.png)

veracioux commented:
Alright, new stuff:
- Added directory fallback scheme as we agreed: `Pictures`, `$HOME`, `/tmp`
- Added a `flameshot config --check` option that reports errors in the config
- Added the GUI version of the above - just introduce an error, open the config window and click the "Show errors" button.

mmahmoudian commented:
Looks great 👍🏼  . The parts that got me really excited were:
1. The CLI that gives nice explanation (although it can be improved but for the time being it is definitely good)
2. The notification that states that the problem is successfully resolved.

⚠️ During my testing there was one type of error that it didn't capture. it considered `#ff7300aaa` a valid color which as far as I know a hex color code should have either of 3,6, or 8 hexadecimal characters.

veracioux commented:
@mmahmoudian I just used Qt's color parser. Turns out that it accepts `#RRRGGGBBB` and `#RRRRGGGGBBBB` as well. I'll manually mark these as invalid. And also they use `#AARRGGBB` instead of `#RRGGBBAA`. Just to verify with you - we want to use the latter, right?

mmahmoudian commented:
> Turns out that it accepts #RRRGGGBBB and #RRRRGGGGBBBB as well. 

🤯  I didn't know that. if QT really accept them and they can turn into a color and they do not break Flameshot, I guess we can keep it in.

mmahmoudian commented:
@veracioux Is it possible to as the QT's color parser to only accept #RRGGBBAA ?

veracioux commented:
> What is your preference?

I agree, I don't expect it to cause any issues with the users. And it's less work for me :)

> Is it possible to as the QT's color parser to only accept #RRGGBBAA ?
I've implemented that case manually.

veracioux commented:
@mmahmoudian @borgmanJeremy 
I think I have implemented everything we discussed, except for file logging. I would like to leave that for another time, when we figure out a proper application interface (I don't want to implement a half-cocked solution only to have to change it later). I have been working on some ideas to create a common interface for sending notifications, CLI messages (stdout/err) and file logging, and will be sending a PR for that sometime.

Please gloss over the TODO entries at the top and let me know if I missed something.

@mmahmoudian If you have ideas on how to improve the `config --check` option, please share.

I say we do some more testing, and then release it into the wild and see how it performs :wink:

veracioux commented:
> @veracioux Is it possible to as the QT's color parser to only accept #RRGGBBAA ?
> 
> This is because then the behavior of Flameshot becomes deterministic and we can handle bug-reports easier

Wait, you mean we should only accept `#RRGGBBAA` out of all possible ways to specify the color? In other words, drop `#RGB`, `RRGGBB`, and the others?

mmahmoudian commented:
@veracioux I want what ever that is easiest for you now and the least confusing for the user and maintainers in the long run. I have no preference here. Basically I'm totally fine with whatever QT accepts. But we should probably include a link to the QT color parser documentation of the user has a wrong setting.

What is your preference?

veracioux commented:
@mmahmoudian I say we allow #RRGGBB, #RRGGBBAA and by color name (e.g. red). And ban the other formats. These are the most common, most intuitive and they cover basically every use case. I wouldn't want to redirect the user to the Qt doc in any circumstance.

veracioux commented:
@mmahmoudian What else is left to do?

mmahmoudian commented:
Imo this is very much ready 👍🏼
I have one suggestion but that might be too much trouble as it is just for aesthetic, so I just leave it here:

The `flameshot config -check` can return the error in a colored form. For example the name of the key in yellow and the error after the : in red.

mmahmoudian commented:
Also, is it possible to have comment in the ini file? For example leading the line with # makes that line comment?

veracioux commented:
> The `flameshot config -check` can return the error in a colored form. For example the name of the key in yellow and the error after the : in red.

Because I use the same code for both CLI and GUI error reporting, I can't do that without some trouble. Basically: elegant CLI interface => inelegant code in this case.

> Also, is it possible to have comment in the ini file? For example leading the line with # makes that line comment?

Problem is that QSettings removes comments when it writes settings to the file. I think it's enough to have the example config which is well documented as it is. Although I think it would be good to install the example config in a standard location like `/usr/share/flameshot` or `/usr/share/doc/flameshot` (on linux at least). I'm not volunteering to do that since I'm not currently in the mood to dig into makefiles and the like.

veracioux commented:
Alright, then we are waiting for your final word @borgmanJeremy

borgmanJeremy commented:
Thanks for the hardwork! merging it now.

mmahmoudian commented:
Thanks @veracioux for working on this 👍🏼 🚀

veracioux commented:
@borgmanJeremy @mmahmoudian Thank you! It was an absolute pleasure.
